File: binary_tree_to_bst.java
//brute force
class Solution
{
    // The given root is the root of the Binary Tree
    // Return the root of the generated BST
    void helper(Node root, ArrayList<Integer> preorder){
        if(root == null) return;
        preorder.add(root.data);
        helper(root.left, preorder);
        helper(root.right, preorder);
    }

    void insert(Node node, int data){
        if(node.left == null && node.data > data){
            node.left = new Node(data);
        }
        if(node.right == null && node.data < data){
            node.right = new Node(data);
        }
        if(node.data > data){
            insert(node.left, data);
        }
        else if(node.data < data){
            insert(node.right, data);
        }
    }

    Node binaryTreeToBST(Node root)
    {
        if(root == null) return null;
        ArrayList<Integer> preorder = new ArrayList<>();
        helper(root, preorder);
        Collections.sort(preorder);
        Node node = new Node(preorder.get(0));
        for(int i = 1 ; i<preorder.size() ; i++){
            insert(node, preorder.get(i));
        }
        return node;
    }
}

//optimised way to do that

File: binary_tree_to_bst.cpp
class Solution{
public:
    void inOrder(Node* root, vector<int>& inorder){
        if(root == nullptr){
            return;
        }
        inOrder(root->left, inorder);
        inorder.push_back(root->data);
        inOrder(root->right, inorder);
    }
    Node* constructBST(vector<int> inorder, int start, int end){
        if(start > end){
            return nullptr;
        }
        int mid = (start + end)/2;
        Node* node = new Node(inorder[mid]);
        node->left = constructBST(inorder, start, mid-1);
        node->right = constructBST(inorder, mid+1, end);
        return node;
    }
    Node *binaryTreeToBST (Node *root)
    {
        vector<int> inorder;
        inOrder(root, inorder);
        sort(inorder.begin() , inorder.end());
        return constructBST(inorder, 0 , inorder.size()-1);
    }
};